cpp/src/arrow/compute/kernels/vector_selection.cc

```diff
@@ -0,0 +1,1816 @@
+        null_selection_(null_selection) {
+    ArrayData* out_arr = out_datum->mutable_array();
+    out_is_valid_ = out_arr->buffers[0]->mutable_data();
+    out_data_ = reinterpret_cast<T*>(out_arr->buffers[1]->mutable_data());
+    out_offset_ = out_arr->offset;
+    out_length_ = out_arr->length;
+    out_position_ = 0;
+  }
+
+  void ExecNonNull() {
+    // The result is all not-null
+    BitUtil::SetBitsTo(out_is_valid_, out_offset_ + out_position_, out_length_, true);
+
+    // Fast filter when values and filter are not null
+    // Bit counters used for both null_selection behaviors
+    BitBlockCounter filter_counter(filter_data_, filter_offset_, values_length_);
+
+    int64_t in_position = 0;
+    BitBlockCount current_block = filter_counter.NextWord();
+    while (in_position < values_length_) {
+      if (current_block.AllSet()) {
+        int64_t run_length = 0;
+        // If we've found a all-true block, then we scan forward until we find
+        // a block that has some false values (or we reach the end
+        while (current_block.length > 0 && current_block.AllSet()) {
+          run_length += current_block.length;
+          current_block = filter_counter.NextWord();
+        }
+        WriteValueSegment(in_position, run_length);
+        in_position += run_length;
+      } else if (current_block.NoneSet()) {
+        // Nothing selected
+        in_position += current_block.length;
+        current_block = filter_counter.NextWord();
+      } else {
+        // Some values selected
+        for (int64_t i = 0; i < current_block.length; ++i) {
+          if (BitUtil::GetBit(filter_data_, filter_offset_ + in_position)) {
+            WriteValue(in_position);
+          }
+          ++in_position;
+        }
+        current_block = filter_counter.NextWord();
+      }
+    }
+  }
+
+  void Exec() {
+    if (filter_null_count_ == 0 && values_null_count_ == 0) {
+      return ExecNonNull();
+    }
+
+    // Bit counters used for both null_selection behaviors
+    DropNullCounter drop_null_counter(filter_is_valid_, filter_data_, filter_offset_,
+                                      values_length_);
+    OptionalBitBlockCounter data_counter(values_is_valid_, values_offset_,
+                                         values_length_);
+    OptionalBitBlockCounter filter_valid_counter(filter_is_valid_, filter_offset_,
+                                                 values_length_);
+
+    auto WriteNotNull = [&](int64_t index) {
+      BitUtil::SetBit(out_is_valid_, out_offset_ + out_position_);
+      // Increments out_position_
+      WriteValue(index);
+    };
+
+    auto WriteMaybeNull = [&](int64_t index) {
+      BitUtil::SetBitTo(out_is_valid_, out_offset_ + out_position_,
+                        BitUtil::GetBit(values_is_valid_, values_offset_ + index));
+      // Increments out_position_
+      WriteValue(index);
+    };
+
+    int64_t in_position = 0;
+    while (in_position < values_length_) {
```

Comment:
This is very hard to read and review, the nesting is deep and tracking of loops invariant is almost impossible.

Reply:
Well, I'm not going to refactor it in this PR. Getting it to be correct and fast was time consuming.